pkg/assembler/assembler.go

```diff
@@ -140,3 +146,118 @@ func (g *Graph) AppendGraph(gs ...Graph) {
 
 // AssemblerInput represents the inputs to add to the graph
 type AssemblerInput = Graph
+
+// UnmarshalJSON deserializes back to graph struct with GuacNode and GuacEdge
+func (g *Graph) UnmarshalJSON(b []byte) error {
+
+	var objMap map[string]*json.RawMessage
+	err := json.Unmarshal(b, &objMap)
+	if err != nil {
+		return err
+	}
+
+	var rawMessagesForGuacNodes []*json.RawMessage
+	err = json.Unmarshal(*objMap["Nodes"], &rawMessagesForGuacNodes)
+	if err != nil {
+		return err
+	}
+
+	var rawMessagesForGuacEdges []*json.RawMessage
+	err = json.Unmarshal(*objMap["Edges"], &rawMessagesForGuacEdges)
+	if err != nil {
+		return err
+	}
+
+	g.Nodes = make([]GuacNode, len(rawMessagesForGuacNodes))
+	g.Edges = make([]GuacEdge, len(rawMessagesForGuacEdges))
+
+	foundNodes, err := unmarshalNodeType(rawMessagesForGuacNodes, map[string]reflect.Type{
+		ArtifactNodeType:      reflect.TypeOf(ArtifactNode{}),
+		PackageNodeType:       reflect.TypeOf(PackageNode{}),
+		IdentityNodeType:      reflect.TypeOf(IdentityNode{}),
+		AttestationNodeType:   reflect.TypeOf(AttestationNode{}),
+		BuilderNodeType:       reflect.TypeOf(BuilderNode{}),
+		MetadataNodeType:      reflect.TypeOf(MetadataNode{}),
+		VulnerabilityNodeType: reflect.TypeOf(VulnerabilityNode{}),
+	})
+	if err != nil {
+		return err
+	}
+
+	foundEdges, err := unmarshalEdgeType(rawMessagesForGuacEdges, map[string]reflect.Type{
+		IdentityForEdgeType:    reflect.TypeOf(IdentityForEdge{}),
+		AttestationForEdgeType: reflect.TypeOf(AttestationForEdge{}),
+		BuiltByEdgeType:        reflect.TypeOf(BuiltByEdge{}),
+		DependsOnEdgeType:      reflect.TypeOf(DependsOnEdge{}),
+		ContainsEdgeType:       reflect.TypeOf(ContainsEdge{}),
+		MetadataForEdgeType:    reflect.TypeOf(MetadataForEdge{}),
+		VulnerableEdgeType:     reflect.TypeOf(VulnerableEdge{}),
+	})
+	if err != nil {
+		return err
+	}
+
+	g.Nodes = foundNodes
+	g.Edges = foundEdges
+
+	return nil
+}
+
+func unmarshalNodeType(rawMessagesForGuacNodes []*json.RawMessage, customTypes map[string]reflect.Type) ([]GuacNode, error) {
```

Comment:
i'm wondering since we do control both ends here, would it help to define a serialization format struct instead? May make things a bit cleaner. Thoughts?

Maybe something like

```
type GuacNodeSerialization struct {
    Type string `json:"type"`
    MetdataNode MetadataNode `json:"metadataNode"`
    ArtifactNode ArtifactNode `json:"artifactNode"`
...
}
```

or potentially just on the graph level (although I think it probably is helpful to have it on the node/edge level as well.

Reply:
hmm not sure I am following completely. Can you elaborate a bit more?

Reply:
going to try my best to describe it in prose, but may fail terribly ...

The thought is having `func (g Graph) MarshalJSON`  to take the list of Guac nodes and then based on a switch statement on their type, populate a new struct with a list of each type of node.. So the serializable wire format of the `Graph` is different from the one in golang. 

So Marshal would:
- convert Graph to GuacNodeSerialization
- Serialize GuacNodeSerialization struct to JSON

Unmarshal would:
- Unmarshal JSON to GuacNodeSerialization struct
- convert GuacNodeSerialization to Graph

Hope this articulates the idea better.. will be happy to find some time to hop on the call for more clarity

Reply:
I'd say let's not focus on this, since the GraphQL interface will also serialize stuff that goes to the DB

Reply:
@mihaimaruseac let's chat about this. I can re-open the PR.